File: code/naive-bayes-classifier/python/classifier.py
import re
import math
from collections import Counter, defaultdict


class Classifier:
    """
    A simple implementation of the Naive-Bayes classifier
    """

    training_set: dict
    training_labels: set

    def __init__(self, training_set: dict):
        self.init_training_set(training_set)

    def tokenize(self, msg: str) -> set:
        """
        Finds all words from the specified message and returns a set object with their words
        :param msg:
        :return:
        """

        return set(re.findall(r'\w+', msg.lower()))

    def init_training_set(self, data: dict):
        """
        Initializes the specified data as a training set
        :param data:
        :return:
        """

        self.training_labels = set(data.keys())

        words_count = dict()
        total = Counter()

        for key, cluster in data.items():
            for message in cluster:
                for word in self.tokenize(message):
                    if word not in words_count:
                        words_count[word] = {key: 0 for key in self.training_labels}

                    words_count[word][key] += 1
                    total[key] += 1

        training_set = defaultdict(lambda: defaultdict(float))
        k = 0.5

        for word, clusters in words_count.items():
            for key in clusters:
                training_set[word][key] = (k + clusters[key]) / (2 * k + total[key])

        self.training_set = training_set

    def check(self, msg: str) -> dict:
        """
        Checks the specified message with a training set and returns a dictionary with probabilities
        :param msg:
        :return:
        """

        tokens = self.tokenize(msg)
        values = [0] * len(self.training_labels)

        for word, clusters in self.training_set.items():
            for i, key in enumerate(clusters.keys()):
                if word in tokens:
                    values[i] += math.log(clusters[key])

                else:
                    values[i] += math.log(1 - clusters[key])

        values = list(map(math.exp, values))
        sum_of_values = sum(values)
        result = dict()

        for i, key in enumerate(self.training_labels):
            result[key] = values[i] / sum_of_values

        return result

File: code/naive-bayes-classifier/rust/src/classifier.rs
/// A simple implementation of the Naive-Bayes classifier

use regex::Regex;
use std::collections::{HashMap, HashSet};
use crate::training_set::TrainingSet;

#[derive(Debug)]
pub struct Classifier<'a> {
    labels: Vec<&'a str>,
    training_set: HashMap<String, HashMap<&'a str, f32>>
}

impl <'a> Classifier<'a> {
    /// Finds all words from the specified message and returns a set object with their words
    pub fn tokenize(msg: &str) -> HashSet<String> {
        let re = Regex::new(r"\p{L}+").unwrap();

        re.find_iter(msg).fold(HashSet::new(), |mut set, el| {
            set.insert(el.as_str().to_owned());
            set
        })
    }

    /// Initializes the specified data as a training set
    pub fn init_training_set(data: &'a TrainingSet, k: Option<f32>) -> Self {
        let nk = k.or(Some(0.5)).unwrap();

        let labels: Vec<_> = data.store
            .keys()
            .map(|el| el.as_str())
            .collect();

        let mut total: HashMap<_, u32> = HashMap::new();
        let mut training_set = HashMap::new();

        for (label, messages) in &data.store {
            for msg in messages {
                for word in Classifier::tokenize(&msg) {
                    let clusters = training_set.entry(word).or_insert_with(|| {
                        let mut values = HashMap::new();

                        for label in &labels {
                            values.insert(label.to_owned(), 0.0);
                        }

                        values
                    });

                    *clusters.get_mut(label.as_str()).unwrap() += 1.0;
                    *total.entry(label.as_str()).or_default() += 1;
                }
            }
        }

        for (_, labels) in &mut training_set {
            for (label, value) in labels {
                *value = (nk + *value) / (2.0 * nk + *total.get(label).unwrap() as f32)
            }
        }

        Classifier {
            labels,
            training_set
        }
    }

    /// Checks the specified message with a training set and returns a dictionary with probabilities
    pub fn check(&self, msg: &str) -> HashMap<&str, f32> {
        let mut values = vec![0f32; self.labels.len()];
        let tokens = Classifier::tokenize(msg);

        for (word, clusters) in &self.training_set {
            for (i, label) in self.labels.iter().enumerate() {
                if tokens.contains(word) {
                    values[i] += clusters.get(label).unwrap().ln();

                } else {
                    values[i] += (1.0 - *clusters.get(label).unwrap()).ln();
                }
            }
        }

        values = values.iter().map(|el| el.exp()).collect();

        let sum_of_values: f32 = values.iter().sum();
        let mut result = HashMap::new();

        for (i, label) in self.labels.iter().enumerate() {
            result.insert(*label, values[i] / sum_of_values);
        }

        result
    }
}
